Fix OAI Identify response

drbauder opened the pull request:
Redmine: https://projects.artefactual.com/issues/9395

I noticed that the XML returned by the OAI Identify response does not validate. Validation would fail because of the format of the repositoryIdentifer and sampleIdentifier elements. This patch fixes the output of those elements and ensures validation succeeds.

This is my first pull request here, so any and all feedback is welcome.

Refer to the following:
- http://www.openarchives.org/OAI/2.0/openarchivesprotocol.htm#Identify, for information on the Identify response
- http://www.openarchives.org/OAI/2.0/guidelines.htm, section 3.1, for information on the description container
- http://www.openarchives.org/OAI/2.0/guidelines-oai-identifier.htm, for specifics on the oai-identifier format

drbauder committed:
Fix to bring OAI Identify response in line with OAI specs

According to the OAI specs <repositoryIdentifier> element is not meant
to distinguish repositories; it is a holdover from OAI 1.0/1.1 and it
should contain the namespace-identifier (domain name).

Updated XML response, so that it passes schema validation, by
- removing repository code from the repositoryIdentifier element;
- and adding 'oai:' scheme prefix to the sampleIdentifier element.

drbauder committed:
Add name/email as author credit

sevein commented:
@drbauder, can you get rid of e8ff5dc? You should probably use `git-rebase` instead to keep your branch up to date. Thank you!

drbauder commented:
@sevein, there you go! Thanks for the tip.

I'm working on getting a Contributor's Agreement signed for our institution, but hopefully that won't hold up any other review you want to do.

sevein commented:
That looks much better, thank you.

drbauder committed:
Code formatting

Convert leading tabs to spaces and remove space around dot operator.

drbauder commented:
The only one that hurts my soul is the lack of whitespace around an operator ;-)

sevein commented:
On Wed, Jan 27, 2016 at 8:26 AM, Damian Bauder notifications@github.com
wrote:

> The only one that hurts my soul is the lack of whitespace around an
> operator ;-)
> 
> We borrowed that habit from Symfony 1.x, see https://git.io/vz9r7.

## 

Jesús García Crespo,
Software Engineer, Artefactual Systems Inc.
http://www.artefactual.com | +1.604.527.2056

drbauder committed:
Use siteBaseUrl setting and simplify code

getOaiNamespaceIdentifer() now uses the siteBaseUrl setting.

Also, call getOaiNamespaceIdentifier() or getRepositoryIdentifer, as
appropriate, instead of retrieving the host or siteBaseUrl setting in
multiple places.

sevein commented:
@drbauder, thank you so much! I've asked @fiver-watson to test it.

sevein commented:
@fiver-watson and I have been testing this locally and it seems to work as expected. Something that we've noticed though is that `<baseURL>` is populated using [a different function in the QubitOai class (getBaseUrl)](https://github.com/drbauder/atom/blob/fix-oai-identify-response/lib/QubitOai.class.php#L168-L181). How do you feel about updating it to take advantage of the `siteBaseUrl` setting? That function might as well be used by `getOaiNamespaceIdentifier`.

drbauder commented:
@sevein, I'm not sure I'm clear. Are you suggesting that we update `getBaseUrl` to make use of the `siteBaseUrl` setting, and then call `getBaseUrl` in `getOaiNamespaceIdentifier` as well as in `<baseUrl>`? That makes sense to me for a number of reasons, but before I get too far into it, I want to make sure I'm understanding your comment!

fiver-watson commented:
Hi @drbauder , I'll let @sevein respond re: the technical aspects, but from my perspective: 

Your changes looked great! While I was testing I happened to notice that the <baseURL> value in the Indentify response was not populating correctly. I asked @sevein to check if this was a previously unnoticed bug or due to a change that you'd made. It's nothing you did, but he noticed that the way the Base URL is being fetched for the <baseURL> output in the response is using a different function than the one you used in your changes to the repositoryidentifier... which works! As far as I understand, the proposal was to see if you were willing to address that issue as well, and populate the <baseURL> value using the `siteBaseUrl` method/function/whatever... beyond this, I wasn't clear if @sevein was also suggesting changes to `getOaiNamespaceIdentifier`; I will let him clarify. 

If you don't have the time to take this on, I think your PR looks fine - hopefully we can fix the <baseURL> issue I noticed at a future date. If you were willing though, we'd be grateful!

drbauder commented:
Thanks, @fiver-Watson. I don't mind adding to the changes. It is probably a pretty easy fix to throw in. The one thing I'm not clear on, that maybe @sevein can fill me in on, is whether `SERVER['SERVER_NAME']` will contain the proper domain in a situation with a proxy, which was why he suggested using `siteBaseUrl` in the first place. If `SERVER['SERVER_NAME']` actually plays nice with proxies, it might make more sense to use `getBaseUrl` (almost) as-is. It returns a nicely formatted URL that parse_url should get along with quite well.

I say "almost", above, because I noticed a tiny bug in `getBaseUrl` that will return http as the scheme while adding 443 as a port, which will result in a 400 response code should anything try to use it.

Either way, I can drop in whatever fixes are necessary once @sevein fills me in.

sevein commented:
I'd prefer to relay in `siteBaseUrl` because it'd be more predictable for AtoM users that are already familiar with that setting. IIRC, `siteBaseUrl` had to be added because `$_SERVER` is not available in the command line environment, which is frequently used by AtoM users for import/export purposes, e.g. `siteBaseUrl` is used in EAD export, which is a functionality also exposed in the command line.
It'd be probably simpler if we rely on `siteBaseUrl` everywhere rather than implementing two different methods through the application. Also, because `siteBaseUrl` is manually defined by the user, proxies won't ever be a problem. Does that make sense?

drbauder committed:
Update getBaseUrl to use siteBaseUrl and streamline code

Modified getBaseUrl to use the user-defined siteBaseUrl setting, in
conjunction with $_SERVER values. Also, added a helper function to
reliably parse hosts from URLs since parse_url() is slightly broken in
that regard.

drbauder commented:
@sevein and @fiver-watson, I made a few changes. Please take a look!

sevein commented:
LGTM! Thank you so much.
We'll merge soon. @fiver-watson, do you want to give it another test to verify?

drbauder committed:
Use isset() instead of '== null'

drbauder commented:
You are very welcome! Thanks to you both for the review. Hopefully everything works this time through.

sevein commented:
Merged and squashed @ 712ce7b
